### ISSUE TITLE
Add Content-Based Auto-Tagging for Elasticsearch

### ISSUE THREAD

**Issue by jwilleke** [jwilleke/ngdpbase#507]
## Background

During a conversation exploring Elasticsearch tagging strategies and NLP-based auto-tagging, we identified an opportunity to automatically generate tags from content rather than requiring manual tagging.

## Conversation Summary

We explored two fundamental approaches to content-based tagging:

1. **Unsupervised extraction** — library reads content and freely generates keywords (e.g. RAKE, TF-IDF, KeyBERT-style)
2. **Zero-shot classification** — provide a predefined list of allowed tags; library picks best matches from that list
3. **Hybrid** — extract freely, then filter/validate against an approved taxonomy

For Elasticsearch specifically, the **"pick from a list"** approach is recommended because aggregations and filters depend on consistent, predictable tag values. Free extraction produces too much variation (e.g. "Kubernetes", "k8s", "kube" as three different tags).

## TypeScript/Node.js Library Options

Since ngdpbase is a TypeScript project, Python-first libraries (KeyBERT, spaCy) are less ideal. Best native options:

| Library | Approach | License | Notes |
|---|---|---|---|
| **Transformers.js** (`@xenova/transformers`) | Zero-shot classification via ONNX | Apache 2.0 | Best accuracy, runs locally, no Python |
| **WinkNLP** (`wink-nlp`) | NER + POS tagging + keyword extraction | MIT | Full NLP pipeline in Node |
| **retext-keywords** | Statistical keyword extraction | MIT | Fully typed TypeScript, unified ecosystem |
| **keyword-extractor** | Stopword removal | MIT | Simple, fast, no ML |

### Phase 1 — Define Tag Taxonomy
- Create a controlled vocabulary of approved tags relevant to the domain
- Store as a config file (e.g. `config/tags.ts`)
- Tags should map to Elasticsearch field values used in aggregations/filters

### Phase 2 — Tagging Service
- Implement a `TaggingService` that accepts document content and returns matched tags
- Use **Transformers.js zero-shot classification** against the approved tag list
- Fallback to **retext-keywords** for unsupervised extraction when no tags match

### Phase 3 — Elasticsearch Integration
- Run tagging pipeline before indexing documents
- Store tags as `keyword` type array field in Elasticsearch mapping
- Use `terms` aggregation for faceted navigation / tag clouds

### Phase 4 — Evaluation
- Compare auto-generated tags vs manually assigned tags on a sample set
- Tune confidence threshold for zero-shot classifier
- Consider hybrid: auto-tag + human review workflow

## References
- [Transformers.js](https://github.com/xenova/transformers.js)
- [WinkNLP](https://winkjs.org/wink-nlp/)
- [retext-keywords](https://github.com/retextjs/retext-keywords)

**Comment by jwilleke** [comment]
## Architecture Update

After further discussion, the actual pipeline is:

```
Files on NAS → sist2 (Linux server) → Elasticsearch → ngdpbase (TypeScript) → Users
```

This changes where tagging should happen and which libraries to use.

## Where Tagging Should Occur

### At Index Time — sist2/Linux (Recommended for file content)
- sist2 already reads the raw file content — this is the right place to tag
- On Linux, the full Python Hugging Face Transformers library is available
- Tags get stored in Elasticsearch alongside the document
- ngdpbase just reads pre-tagged data — no tagging work needed on the read side

### In ngdpbase — TypeScript (For app-level tagging only)
- Only needed for content that originates inside ngdpbase itself
- User input, notes, app-generated content etc.
- Would use Transformers.js for this

## Hugging Face Transformers vs Transformers.js

These are the same thing — different runtimes for the same models:

| | Hugging Face Transformers | Transformers.js |
|---|---|---|
| Language | Python | TypeScript/JS |
| Where | Linux/sist2 server | Node.js (ngdpbase) |
| Models | Hugging Face Hub | Same models, converted to ONNX |
| GPU support | Yes | Rarely |
| Best for | sist2 pipeline (file tagging) | ngdpbase app-level tagging |

## Open Source & Royalty-Free Confirmation

Both libraries are fully open source and royalty-free:

- **Hugging Face Transformers** — Apache 2.0
- **Transformers.js** — Apache 2.0
- **Most popular models** (BERT, RoBERTa, DistilBERT, all-MiniLM) — Apache 2.0 or MIT

No per-call fees, no API keys, runs entirely on your own hardware. Significant advantage over commercial options like OpenAI or Google NLP API.

## ML vs Non-ML Clarification

| Approach | ML? | Training Needed? | Recommended? |
|---|---|---|---|
| RAKE / keyword-extractor / retext-keywords | No — statistical only | No | Prototype only |
| Hugging Face Transformers (Python) | Yes — pre-trained | No | ✅ sist2 pipeline |
| Transformers.js | Yes — pre-trained | No | ✅ ngdpbase app tagging |
| Custom spaCy / scikit-learn model | Yes — you train it | Yes | Only if domain-specific accuracy needed |

Hugging Face Transformers gives full ML power with zero training required — pre-trained models already understand language from training on billions of words.

## Next Steps

- Clarify all tagging use cases needed within ngdpbase beyond file content
- Determine if sist2 supports a plugin/hook for running a tagging script at index time
- Define the approved tag taxonomy (controlled vocabulary) for zero-shot classification

**Comment by jwilleke** [comment]
## Implementation — Temp Instance (2026-04-13)

Phase 1 (taxonomy) is complete via #424 (CatalogManager with 9 subject terms: geology, medicine, meteorology, oceanography, psychology, sociology, anthropology, economics, political-science).

**Phase 2 — TaggingService (statistical, no ML)** implemented in `src/utils/TaggingService.ts`:
- Token-overlap matching: each vocabulary term's tokens are checked against tokenized page content
- Threshold: 0.5 (half of term's tokens must appear in content)
- Subject-category terms only — workflow-status (draft/review/published) intentionally excluded
- 17 unit tests pass

**Phase 3 — ES integration** in `ElasticsearchSearchProvider._pageToDoc()`:
- TaggingService loaded at initialize() if `ngdpbase.search.provider.elasticsearch.autotagging.enabled: true`
- Matched terms merged into `systemKeywords` (deduped with existing manual tags)
- Disabled by default; enabled per-instance via custom config

**Temp instance results (port 3001):**
- 103/103 pages indexed
- 4 pages got auto-tagged systemKeywords (geology, medicine, oceanography, etc.)
- False positives on documentation pages that list vocabulary terms in examples — expected behavior

Commits: `f6c07de7`, `6828a25d`

**Next steps for promotion:**
- Evaluate tag quality on ve-geology (which has actual geological content)
- Consider lowering threshold or adding label-only tokens for better recall
- ML upgrade path: swap TaggingService for Transformers.js zero-shot classifier if needed

**Comment by jwilleke** [comment]
## Status check (2026-04-30)

### Implemented and merged

- **Phase 1 (taxonomy)** — delivered via #424. `CatalogManager` ships with 9 subject terms (geology, medicine, meteorology, oceanography, psychology, sociology, anthropology, economics, political-science).
- **Phase 2 (TaggingService)** — `src/utils/TaggingService.ts` is in place. Statistical token-overlap matcher; threshold 0.5; subject-category terms only (workflow-status excluded). 17 unit tests passing.
- **Phase 3 (Elasticsearch integration)** — `ElasticsearchSearchProvider` imports `TaggingService` (`src/providers/ElasticsearchSearchProvider.ts:41,115,146,151`); auto-tagged terms merge into `systemKeywords` at index time. Config key `ngdpbase.search.provider.elasticsearch.autotagging.enabled` (default `false`) at `config/app-default-config.json:104`.

Commits: `6828a25d`, `5b388649`, `f6c07de7`.

### Phase 4 — not done

The original issue body called Phase 4 "Compare auto-generated tags vs manually assigned tags on a sample set; tune confidence threshold; consider hybrid auto-tag + human review." Outstanding:

- Quality evaluation on real content. Temp-instance run was 103/103 indexed, only 4 auto-tagged, with admitted false positives on doc pages that list vocabulary terms in examples.
- Threshold tuning or label-only tokens for better recall.
- ML upgrade path (Transformers.js zero-shot classifier) — optional.

### Recommendation

The code is shipped and configurable. Phase 4 evaluation is operational work that arguably belongs in a separate follow-up issue rather than keeping this one open indefinitely. Marking `in review` so you can decide:

1. Close this issue and open a new `[FEATURE] Auto-tagging quality evaluation` for Phase 4 work, or
2. Keep it open and complete Phase 4 here.

**Comment by jwilleke** [comment]
## Cross-link: [es-enrichment-pipeline](https://github.com/jwilleke/es-enrichment-pipeline)

Surfacing this for the Phase 4 "ML upgrade path" decision.

`es-enrichment-pipeline` is a standalone TypeScript service that enriches Elasticsearch indexes out-of-band, using local Ollama models. Its description lists **auto-tags** as a goal alongside embeddings (today the README only documents embeddings — vector enrichment via `nomic-embed-text` written to a `<source>-embeddings` mirror index).

### Why it's relevant to #507

The original Phase 4 plan was "swap TaggingService for Transformers.js zero-shot classifier" — ML running **in-process** inside ngdpbase. This pipeline is the same outcome via a different architectural seam:

- ML stays **out-of-process** (Ollama daemon, separate from ngdpbase).
- Enrichment writes tags into a sibling ES index, never touching the source index.
- ngdpbase reads enriched tags at query time, no per-request inference, no Transformers.js memory cost in the ngdpbase daemon.

For the small-org / non-profit deployment target (low memory headroom), this is a strictly better architecture than embedding ML inside ngdpbase. It also matches the "indexer at the ES layer" pipeline that already exists for sist2 → ES.

### Caveat

The pipeline currently only delivers **embeddings**. Auto-tag generation is listed in the repo's description but not yet implemented per the README. So this is a *future* Phase 4 path, not today.

### Suggested action

Once `es-enrichment-pipeline` adds tag generation (writing into a known mirror field), revisit this issue: ngdpbase's `TaggingService` could become a thin reader of those mirror-index tags, and the in-process tagging path could be retired. No code change needed in ngdpbase until then.

**Comment by jwilleke** [comment]
Phases 1-3 verified shipped:

- `src/utils/TaggingService.ts` (Phase 2 — token-overlap matcher, threshold 0.5, 17 unit tests).
- `ElasticsearchSearchProvider` integrates TaggingService at index time (4 references); gated by `ngdpbase.search.provider.elasticsearch.autotagging.enabled` (default false).
- Phase 1 (taxonomy) provided by #424 CatalogManager.

Phase 4 (quality evaluation: compare auto-tags vs manual on real content, tune threshold, optional ML upgrade via Transformers.js or via the [es-enrichment-pipeline](https://github.com/jwilleke/es-enrichment-pipeline) cross-link) is operational work — better tracked as a separate issue if you want it pursued. Closing this since all the code-shipping phases are done.
